### ISSUE TITLE
[vLLM] GPT-2 GPU E2E fails on final_logits dtype mismatch

### ISSUE THREAD

**Issue by Samfisheryu** [ProjectDMX/DMI#97]
## Summary

The vLLM identical GPU E2E in the DMI checkout completes inference and compares 306 captures, but two GPT-2 `final_logits` checks fail:

```text
304 passed, 2 failed
0/final_logits_T5_6: dtype mismatch: torch.float32 vs torch.float16
0/final_logits_T6_7: dtype mismatch: torch.float32 vs torch.float16
```

Original vs reference and original vs monitored logprobs are both bitwise identical. This currently blocks treating the vLLM GPU E2E as a clean PR/release gate.

## Tested revision and environment

- DMI PR #95 head: `0996fe3c4cd1c8777a8f27923e20fda9a0450cb2`
- DMI-vLLM-Integration submodule: `afc0896451ce2a5be080f3d8cd636e45ff681076`
- vLLM: `0.27.1`
- PyTorch: `2.13.0+cu130`
- Python: `3.10.12`
- GPU: RTX 4090
- CUDA toolkit/runtime used for the native build: 13.0
- ClickHouse: 25.12

## Reproduction

Install DMI and the vLLM integration, use official vLLM 0.27.1 with a supported Transformers v5 dependency, build the DMI native backend against one matching CUDA toolkit/runtime, and start ClickHouse.

```bash
cd third_party/vllm-integration

CUDA_VISIBLE_DEVICES=1 \
VLLM_USE_FLASHINFER_SAMPLER=0 \
E2E_MODEL=gpt2 \
E2E_NUM_PROMPTS=1 \
E2E_MAX_NEW_TOKENS=2 \
E2E_DTYPE=float16 \
E2E_ENFORCE_EAGER=1 \
E2E_RING_PAYLOAD_MB=64 \
E2E_RING_PINNED_MB=64 \
E2E_GPU_MEM_UTIL=0.4 \
E2E_KEEP_ARTIFACTS=1 \
E2E_ARTIFACT_DIR=/tmp/dmi-vllm-gpt2-repro \
pytest -q -s tests/test_identical.py::test_vllm_identical
```

## Likely cause

The GPT-2 reference oracle hard-codes its `final_logits` buffer to `torch.float32`:

```python
self._buf_final_logits = _alloc(max_len, vocab_size, dtype=torch.float32)
```

The monitored model captures the logits directly, while the adapter assigns the vLLM model head dtype to the hook spec. With `E2E_DTYPE=float16`, the monitored capture is therefore `float16`.

Relevant files in the integration submodule:

- `tests/oracles/gpt2_ref.py`
- `src/dmi_vllm_integration/models/gpt2.py`
- `src/dmi_vllm_integration/adapter.py`

## Expected result

Reference and monitored captures should follow one explicit dtype contract and all 306 comparisons should pass. Either the reference oracle should preserve the actual logits/head dtype, or monitored capture should explicitly cast if float32 persistence is intentional.

Implementation-level tracking and detailed evidence: ProjectDMX/DMI-vLLM-Integration#11.

Related test-matrix work: #54.

**Comment by XbzOnGit** [comment]
Resolved by ProjectDMX/DMI-vLLM-Integration#14. This was a validation-oracle dtype bug; production capture and inference behavior were unchanged. The exact reported reproduction now passes all 306/306 capture comparisons, with original-vs-reference and original-vs-monitored logprobs both bitwise exact.
